Valid string generates invalid code point exception

### Description
Robolectric tries to convert to code point a string that happens to contain `\u` (not a code point) which has been properly escaped.

Stacktrace:
<details>

```
java.lang.RuntimeException: java.lang.RuntimeException: error parsing build/intermediates/res/merged/debug/values/values.xml
	at org.robolectric.res.ResourceTableFactory.parseResourceFiles(ResourceTableFactory.java:155)
	at org.robolectric.res.ResourceTableFactory.newResourceTable(ResourceTableFactory.java:43)
	at org.robolectric.res.ResourceMerger.buildResourceTable(ResourceMerger.java:22)
	at org.robolectric.RobolectricTestRunner.getAppResourceTable(RobolectricTestRunner.java:450)
	at org.robolectric.RobolectricTestRunner.beforeTest(RobolectricTestRunner.java:288)
	at org.robolectric.internal.SandboxTestRunner$2.evaluate(SandboxTestRunner.java:203)
	at org.robolectric.internal.SandboxTestRunner.runChild(SandboxTestRunner.java:109)
	at org.robolectric.internal.SandboxTestRunner.runChild(SandboxTestRunner.java:36)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.robolectric.internal.SandboxTestRunner$1.evaluate(SandboxTestRunner.java:63)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecuter.runTestClass(JUnitTestClassExecuter.java:114)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecuter.execute(JUnitTestClassExecuter.java:57)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassProcessor.processTestClass(JUnitTestClassProcessor.java:66)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:51)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:32)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:93)
	at com.sun.proxy.$Proxy3.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.processTestClass(TestWorker.java:109)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:147)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:129)
	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:404)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
	at org.gradle.internal.concurrent.StoppableExecutorImpl$1.run(StoppableExecutorImpl.java:46)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.RuntimeException: error parsing build/intermediates/res/merged/debug/values/values.xml
	at org.robolectric.res.StaxDocumentLoader.loadResourceXmlFile(StaxDocumentLoader.java:32)
	at org.robolectric.res.DocumentLoader.loadFile(DocumentLoader.java:36)
	at org.robolectric.res.DocumentLoader.load(DocumentLoader.java:26)
	at org.robolectric.res.ResourceTableFactory.parseResourceFiles(ResourceTableFactory.java:140)
	... 42 more
Caused by: java.lang.IllegalArgumentException: Invalid code point: \usern
	at org.robolectric.res.StringResources.convertCodePoints(StringResources.java:56)
	at org.robolectric.res.StringResources.processStringResources(StringResources.java:31)
	at org.robolectric.res.StaxValueLoader.onEnd(StaxValueLoader.java:33)
	at org.robolectric.res.StaxDocumentLoader.doParse(StaxDocumentLoader.java:66)
	at org.robolectric.res.StaxDocumentLoader.loadResourceXmlFile(StaxDocumentLoader.java:30)
	... 45 more
Caused by: java.lang.NumberFormatException: For input string: "sern"
	at java.lang.NumberFormatException.forInputString(NumberFormatException.java:65)
	at java.lang.Integer.parseInt(Integer.java:580)
	at java.lang.Integer.valueOf(Integer.java:740)
	at org.robolectric.res.StringResources.convertCodePoints(StringResources.java:51)
	... 49 more
```
</details>

### Steps to Reproduce
Add to `strings.xml`:
```
<string name="valid_string">domain\\username</string>
```

Run any Robolectric test.

### Robolectric & Android Version
Robolectric 3.3.2
Default SDK version (no override)

--- a/robolectric-resources/src/main/java/org/robolectric/res/StringResources.java
+++ b/robolectric-resources/src/main/java/org/robolectric/res/StringResources.java
@@ -1,25 +1,11 @@
 package org.robolectric.res;
 
+import com.google.common.annotations.VisibleForTesting;
+import org.robolectric.util.Logger;
+
 public class StringResources {
 
   private static final int CODE_POINT_LENGTH = 4;
-
-  /**
-   * Provides escaping of String resources as described
-   *
-   * <a href="http://developer.android.com/guide/topics/resources/string-resource.html#FormattingAndStyling">here</a>
-   *
-   * @param text Text to escape.
-   * @return Escaped text.
-   */
-  public static String escape(String text) {
-    if (text.length() > 1 && text.charAt(0) == '"' && text.charAt(text.length() - 1) == '"') {
-      text = text.substring(1, text.length() - 1);
-    } else {
-      text = text.replaceAll("\\\\(['\"])", "$1");
-    }
-    return text;
-  }
 
   /**
    * Processes String resource values in the same way real Android does, namely:-
@@ -28,34 +14,86 @@ public class StringResources {
    * 3) Escapes
    */
   public static String processStringResources(String inputValue) {
-    return escape(convertCodePoints(inputValue.trim()
-            .replace("\\n", String.valueOf('\n'))
-            .replace("\\t", String.valueOf('\t'))
-    ));
+    return escape(inputValue.trim());
+  }
+
+  /**
+   * Provides escaping of String resources as described
+   * <p>
+   * <a href="http://developer.android.com/guide/topics/resources/string-resource.html#FormattingAndStyling">here</a>
+   *
+   * @param text Text to escape.
+   * @return Escaped text.
+   */
+  @VisibleForTesting
+  static String escape(String text) {
+    // unwrap double quotes
+    if (text.length() > 1 && text.charAt(0) == '"' && text.charAt(text.length() - 1) == '"') {
+      text = text.substring(1, text.length() - 1);
+    }
+    int i = 0;
+    int length = text.length();
+    StringBuilder result = new StringBuilder(text.length());
+    while (true) {
+      int j = text.indexOf('\\', i);
+      if (j == -1) {
+        result.append(text.substring(i));
+        break;
+      }
+      result.append(text.substring(i, j));
+      if (j == length - 1) {
+        // dangling backslash
+        break;
+      }
+      boolean isUnicodeEscape = false;
+      char escapeCode = text.charAt(j + 1);
+      switch (escapeCode) {
+        case '\'':
+        case '"':
+        case '\\':
+        case '?':
+        case '@':
+        case '#':
+          result.append(escapeCode);
+          break;
+        case 'n':
+          result.append('\n');
+          break;
+        case 't':
+          result.append('\t');
+          break;
+        case 'u':
+          isUnicodeEscape = true;
+          break;
+        default:
+          Logger.strict("Unsupported string resource escape code '%s'", escapeCode);
+      }
+      if (!isUnicodeEscape) {
+        i = j + 2;
+      } else {
+        j += 2;
+        if (length - j < CODE_POINT_LENGTH) {
+          throw new IllegalArgumentException("Too short code point: \\u" + text.substring(j));
+        }
+        String codePoint = text.substring(j, j + CODE_POINT_LENGTH);
+        result.append(extractCodePoint(codePoint));
+        i = j + CODE_POINT_LENGTH;
+      }
+    }
+    return result.toString();
   }
 
   /**
    * Converts code points in a given string to actual characters. This method doesn't handle code
    * points whose char counts are 2. In other words, this method doesn't handle U+10XXXX.
    */
-  private static String convertCodePoints(String src) {
-    String[] tokens = src.split("\\\\u");
-
-    StringBuilder retval = new StringBuilder(tokens[0]);
-    for (int i = 1; i < tokens.length; ++i) {
-      if (tokens[i].length() < CODE_POINT_LENGTH) {
-        throw new IllegalArgumentException("Too short code point: \\u" + tokens[i]);
-      }
-      String codePoint = tokens[i].substring(0, CODE_POINT_LENGTH);
-      try {
-        retval.append(Character.toChars(Integer.valueOf(codePoint, 16)))
-            .append(tokens[i].substring(CODE_POINT_LENGTH));
-      } catch (IllegalArgumentException e) {
-        // This may be caused by NumberFormatException of Integer.valueOf() or
-        // IllegalArgumentException of Character.toChars().
-        throw new IllegalArgumentException("Invalid code point: \\u" + codePoint, e);
-      }
+  private static char[] extractCodePoint(String codePoint) {
+    try {
+      return Character.toChars(Integer.valueOf(codePoint, 16));
+    } catch (IllegalArgumentException e) {
+      // This may be caused by NumberFormatException of Integer.valueOf() or
+      // IllegalArgumentException of Character.toChars().
+      throw new IllegalArgumentException("Invalid code point: \\u" + codePoint, e);
     }
-    return retval.toString();
   }
 }

--- a/robolectric/src/test/java/org/robolectric/res/StringResourcesTest.java
+++ b/robolectric/src/test/java/org/robolectric/res/StringResourcesTest.java
@@ -1,6 +1,9 @@
 package org.robolectric.res;
 
 import org.junit.Test;
+
+import java.util.*;
+
 import static org.assertj.core.api.Assertions.*;
 
 public class StringResourcesTest {
@@ -8,5 +11,57 @@ public class StringResourcesTest {
   public void escape_shouldEscapeStrings() {
     assertThat(StringResources.escape("\"This'll work\"")).isEqualTo("This'll work");
     assertThat(StringResources.escape("This\\'ll also work")).isEqualTo("This'll also work");
+  }
+
+  @Test
+  public void escape_shouldEscapeCodePoints() {
+    Map<String, String> tests = new HashMap<>();
+    tests.put("\\u0031", "1");
+    tests.put("1\\u0032", "12");
+    tests.put("\\u00312", "12");
+    tests.put("1\\u00323", "123");
+    tests.put("\\u005A", "Z");
+    tests.put("\\u005a", "Z");
+
+    for (Map.Entry<String, String> t : tests.entrySet()) {
+      assertThat(StringResources.processStringResources(t.getKey())).isEqualTo(t.getValue());
+    }
+  }
+
+  // Unsupported escape codes should be ignored.
+  @Test
+  public void escape_shouldIgnoreUnsupportedEscapeCodes() {
+    assertThat(StringResources.processStringResources("\\ \\a\\b\\c\\d\\e\\ ")).isEqualTo("");
+  }
+
+  @Test
+  public void escape_shouldSupport() {
+    Map<String, String> tests = new HashMap<>();
+    tests.put("\\\\", "\\");
+    tests.put("domain\\\\username", "domain\\username");
+    for (Map.Entry<String, String> t : tests.entrySet()) {
+      assertThat(StringResources.processStringResources(t.getKey())).isEqualTo(t.getValue());
+    }
+  }
+
+  @Test
+  public void testInvalidCodePoints() {
+    List<String> tests = new LinkedList<>();
+    tests.add("\\u");
+    tests.add("\\u0");
+    tests.add("\\u00");
+    tests.add("\\u004");
+    tests.add("\\uzzzz");
+    tests.add("\\u0zzz");
+    tests.add("\\u00zz");
+    tests.add("\\u000z");
+    for (String t : tests) {
+      try {
+        StringResources.processStringResources(t);
+        fail("expected IllegalArgumentException with test '%s'", t);
+      } catch (IllegalArgumentException expected) {
+        // cool
+      }
+    }
   }
 }

